File: docstrings/tt_os2.py
# -*- coding: utf-8 -*-

# Copyright (c) 2013, Michael Droettboom All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:

# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# The views and conclusions contained in the software and
# documentation are those of the authors and should not be interpreted
# as representing official policies, either expressed or implied, of
# the FreeBSD Project.

from __future__ import print_function, unicode_literals, absolute_import

TT_OS2__init__ = """
Information about the TrueType font, used on OS/2 and Microsoft
Windows.

Note that we now support old Mac fonts which do not include an OS/2
table.  In this case, the `version` field is always set to 0xFFFF.
"""

TT_OS2_version = """
The version of the `TT_OS2` table.

If this table was synthesized from a font that has no OS/2 table, the
version is set to 0xFFFF.
"""

TT_OS2_x_avg_char_width = """
Specifies the arithmetic average of the advance width of all of the 26
lowercase letters of the Latin alphabet and the space character. If
any of the 26 lowercase letters are not present, this parameter should
equal zero.

This parameter is a descriptive attribute of the font that specifies
the spacing of characters used for comparing one font to another for
selection or substitution. For proportionally spaced fonts, this value
is useful in estimating the length for lines of text.
"""

TT_OS2_weight_class = """
A `TT_WEIGHT_CLASS` value.  Indicates the visual weight (degree of
blackness or thickness of strokes) of the characters in the font.
"""

TT_WEIGHT_CLASS = """
|freetypy| Weight values for the `TT_OS2.weight_class` property.

- `ULTRA_LIGHT`

- `EXTRA_LIGHT`

- `LIGHT`

- `SEMI_LIGHT`

- `MEDIUM` (aliased to `NORMAL`)

- `SEMI_BOLD`

- `BOLD`

- `EXTRA_BOLD`

- `ULTRA_BOLD`
"""

TT_OS2_width_class = """
A `TT_WIDTH_CLASS` value.  Indicates a relative change from the normal
aspect ratio (width to height ratio) as specified by a font designer
for the glyphs in a font.

Although every character in a font may have a different numeric aspect
ratio, each character in a font of normal width has a relative aspect
ratio of one. When a new type style is created of a different width
class (either by a font designer or by some automated means) the
relative aspect ratio of the characters in the new font is some
percentage greater or less than those same characters in the normal
font it is this difference that this parameter specifies.
"""

TT_WIDTH_CLASS = """
|freetypy| Width values for the `TT_OS2.width_class` property.

- `ULTRA_CONDENSED`: 50% of normal

- `EXTRA_CONDENSED`: 62.5% of normal

- `CONDENSED`: 75% of normal

- `SEMI_CONDENSED`: 87.5% of normal

- `MEDIUM`: 100% of normal (aliased to `NORMAL`)

- `SEMI_EXPANDED`: 112.5% of normal

- `EXPANDED`: 125% of normal

- `EXTRA_EXPANDED`: 150% of normal

- `ULTRA_EXPANDED`: 200% of normal
"""

TT_OS2_type = """
`FSTYPE` bitflags indicating the licensing restrictions on a font.
"""

TT_OS2_y_subscript_x_size = """
Maps to the em square size of the font being used for a subscript.  If
a font has two recommended sizes for subscripts, e.g., numerics and
other, the numeric sizes should be stressed.  The horizontal font size
specifies a font designer's recommended horizontal font size for
subscript characters associated with this font. If a font does not
include all of the required subscript characters for an application,
and the application can substitute characters by scaling the character
of a font or by substituting characters from another font, this
parameter specifies the recommended em square for those subscript
characters.  For example, if the em square for a font is 2048 and
`y_subscript_x_size` is set to 205, then the horizontal size for a
simulated subscript character would be 1/10th the size of the normal
character.
"""

TT_OS2_y_subscript_y_size = """
Maps to the emHeight of the font being used for a subscript.  If a
font has two recommended sizes for subscripts, e.g. numerics and
other, the numeric sizes should be stressed. The horizontal font size
specifies a font designer's recommendation for horizontal font size of
subscript characters associated with this font. If a font does not
include all of the required subscript characters for an application,
and the application can substitute characters by scaling the
characters in a font or by substituting characters from another font,
this parameter specifies the recommended horizontal EmInc for those
subscript characters.  For example, if the em square for a font is
2048 and `y_subScript_y_size` is set to 205, then the vertical size
for a simulated subscript character would be 1/10th the size of the
normal character.
"""

TT_OS2_y_subscript_x_offset = """
Specifies a font designer's recommended horizontal offset for
subscript characters. It is from the character origin of the font to
the character origin of the subscript's character. If a font does not
include all of the required subscript characters for an application,
and the application can substitute characters, this parameter
specifies the recommended horizontal position from the character
escapement point of the last character before the first subscript
character. For upright characters, this value is usually zero;
however, if the characters of a font have an incline (italic
characters) the reference point for subscript characters is usually
adjusted to compensate for the angle of incline.
"""

TT_OS2_y_subscript_y_offset = """
Specifies a font designer's recommended vertical offset from the
character baseline to the character baseline for subscript characters
associated with this font. Values are expressed as a positive offset
below the character baseline. If a font does not include all of the
required subscript for an application, this parameter specifies the
recommended vertical distance below the character baseline for those
subscript characters.
"""

TT_OS2_y_superscript_x_size = """
Maps to the em square size of the font being used for a subscript.  If
a font has two recommended sizes for subscripts, e.g., numerics and
other, the numeric sizes should be stressed.  The horizontal font size
specifies a font designer's recommended horizontal font size for
superscript characters associated with this font. If a font does not
include all of the required superscript characters for an application,
and the application can substitute characters by scaling the character
of a font or by substituting characters from another font, this
parameter specifies the recommended em square for those superscript
characters.  For example, if the em square for a font is 2048 and
`ySuperScriptXSize` is set to 205, then the horizontal size for a
simulated superscript character would be 1/10th the size of the normal
character.
"""

TT_OS2_y_superscript_y_size = """
Maps to the emHeight of the font being used for a subscript. If a font
has two recommended sizes for subscripts, e.g., numerics and other,
the numeric sizes should be stressed. The vertical font size specifies
a font designer's recommended vertical font size for superscript
characters associated with this font. If a font does not include all
of the required superscript characters for an application, and the
application can substitute characters by scaling the character of a
font or by substituting characters from another font, this parameter
specifies the recommended EmHeight for those superscript characters.
For example, if the em square for a font is 2048 and `ySuperScriptYSize`
is set to 205, then the vertical size for a simulated superscript
character would be 1/10th the size of the normal character.
"""

TT_OS2_y_superscript_x_offset = """
Specifies a font designer's recommended horizontal offset for the
superscript characters associated with this font.  It is from the
character origin to the superscript character's origin. If a font does
not include all of the required superscript characters for an
application, this parameter specifies the recommended horizontal
position from the escapement point of the character before the first
superscript character. For upright characters, this value is usually
zero; however, if the characters of a font have an incline (italic
characters) the reference point for superscript characters is usually
adjusted to compensate for the angle of incline.
"""

TT_OS2_y_superscript_y_offset = """
Specifies a font designer's recommended vertical offset for
superscript characeters.  It is from the character baseline to the
superscript character's baseline associated with this font. Values for
this parameter are expressed as a positive offset above the character
baseline. If a font does not include all of the required superscript
characters for an application, this parameter specifies the
recommended vertical distance above the character baseline for those
superscript characters.
"""

TT_OS2_y_strikeout_size = """
The size of the strikeout line.  This field should normally be the
width of the em-dash for the current font. If the size is one, the
strikeout line will be the line represented by the strikeout position
field. If the value is two, the strikeout line will be the line
represented by the strikeout position and the line immediately above
the strikeout position.
"""

TT_OS2_y_strikeout_position = """
The width of the strikeout line.  Positive values represent distances
above the baseline, while negative values represent distances below
the baseline. A value of zero falls directly on the baseline, while a
value of one falls one pixel above the baseline. The value of
strikeout position should not interfere with the recognition of
standard characters, and therefore should not line up with crossbars
in the font.
"""

TT_OS2_family_class = """
Classifies a font design as to its appearance.  It does not identify
the specific font family, typeface variation, designer, supplier,
size, or metric table differences.  This is the high byte of the
`sFamilyClass` field.

|freetypy| The use of this is limited, so the values are not provided
as enumerations.  See `the TrueType specification
<https://developer.apple.com/fonts/TTRefMan/RM06/Chap6OS2.html>`_ for
more information.
"""

TT_OS2_family_subclass = """
Classifies a font design as to its appearance.  It does not identify
the specific font family, typeface variation, designer, supplier,
size, or metric table differences.  This is the low byte of the
`sFamilyClass` field.

|freetypy| The use of this is limited, so the values are not provided
as enumerations.  See `the TrueType specification
<https://developer.apple.com/fonts/TTRefMan/RM06/Chap6OS2.html>`_ for
more information.
"""

TT_OS2_panose = """
The PANOSE classification number.

|freetypy| The use of this is limited, so the values are not provided
as enumerations.  See `the TrueType specification
<https://developer.apple.com/fonts/TTRefMan/RM06/Chap6OS2.html>`_ for
more information.  A more Pythonic interface may be provided in the
future if needed.

It is a 10-byte string where each byte is as follows:

- 0: Family type
- 1: Serif style
- 2: Weight
- 3: Proportion
- 4: Contrast
- 5: Stroke Variation
- 6: Arm Style
- 7: Letterform
- 8: Midline
- 9: x height
"""

TT_OS2_vend_id = """
Indentifies the font vendor.

It is not the royalty owner of the original artwork but the company
responsible for the marketing and distribution of the typeface that is
being classified. It is reasonable to assume that there will be 6
vendors of ITC Zapf Dingbats for use on desktop platforms in the near
future (if not already). It is also likely that the vendors will have
other inherent benefits in their fonts (more kern pairs, unregularized
data, hand hinted, etc.). This identifier will allow for the correct
vendors type to be used over another, possibly inferior, font
file. These id's are assigned by Microsoft.
"""

TT_OS2_selection = """
A `TT_FS_SELECTION` bitflag.
"""

TT_FS_SELECTION = """
Bitflag concerning the nature of the font patterns.

- `ITALIC`

- `UNDERSCORE`

- `NEGATIVE`

- `OUTLINED`

- `STRIKEOUT`

- `BOLD`
"""

TT_OS2_first_char_index = """
The minimum Unicode index in this font.
"""

TT_OS2_last_char_index = """
The maximum Unicode index in this font.
"""

TT_OS2_typo_ascender = """
The typographic ascender for this font. Remember that this is not the
same as `TT_HoriHeader.ascender`, which Apple defines in a far
different manner.

The suggested usage for `typo_ascender` is that it be used in
conjunction with `TT_Header.units_per_em` to compute a typographically
correct default line spacing. The goal is to free applications from
Macintosh or Windows-specific metrics which are constrained by
backward compatibility requirements. These new metrics, when combined
with the character design widths, will allow applications to lay out
documents in a typographically correct and portable fashion.

For CJK (Chinese, Japanese, and Korean) fonts that are intended to be
used for vertical writing (in addition to horizontal writing), the
required value for `typo_ascender` is that which describes the top of
the of the ideographic em-box. For example, if the ideographic em-box
of the font extends from coordinates 0,-120 to 1000,880 (that is, a
1000x1000 box set 120 design units below the Latin baseline), then the
value of `typo_ascender` must be set to 880. Failing to adhere to these
requirements will result in incorrect vertical layout.
"""

TT_OS2_typo_descender = """
The typographic descender for this font. Remember that this is not the
same as the `TT_HoriHeader.descender`, which Apple defines in a far
different manner.

The suggested usage for `typo_descender` is that it be used in
conjunction with `TT_Header.units_per_em` to compute a typographically
correct default line spacing. The goal is to free applications from
Macintosh or Windows-specific metrics which are constrained by
backward compatability requirements. These new metrics, when combined
with the character design widths, will allow applications to lay out
documents in a typographically correct and portable fashion.

For CJK (Chinese, Japanese, and Korean) fonts that are intended to be
used for vertical writing (in addition to horizontal writing), the
required value for `typo_descender` is that which describes the bottom
of the of the ideographic em-box. For example, if the ideographic
em-box of the font extends from coordinates 0,-120 to 1000,880 (that
is, a 1000x1000 box set 120 design units below the Latin baseline),
then the value of `typo_descender` must be set to -120. Failing to
adhere to these requirements will result in incorrect vertical layout.
"""

TT_OS2_typo_line_gap = """
The typographic line gap for this font. Remember that this is not the
same as `TT_HoriHeader.line_gap`, which Apple defines in a far
different manner.

The suggested usage for `typo_line_gap` is that it be used in
conjunction with `TTHeader.units_per_em` to compute a typographically
correct default line spacing. Typical values average 7-10% of units
per em. The goal is to free applications from Macintosh or
Windows-specific metrics which are constrained by backward
compatability requirements. These new metrics, when combined with the
character design widths, will allow applications to lay out documents
in a typographically correct and portable fashion.  """

TT_OS2_win_ascent = """
The ascender metric for Windows. This, too, is distinct from Apple's
`TT_HoriHeader.ascender` value and from the `typo_ascender`
values. `win_ascent` is computed as the ``y_max`` for all characters
in the Windows ANSI character set. `win_ascent` is used to compute the
Windows font height and default line spacing. For platform 3 encoding
0 fonts, it is the same as `TT_Header.y_max`. Windows will clip the
bitmap of any portion of a glyph that appears above this value. Some
applications use this value to determine default line spacing. This is
strongly discouraged. The typographic ascender, descender and line gap
fields in conjunction with `TT_Header.units_per_em` should be used for
this purpose. Developers should set this field keeping the above
factors in mind.

If any clipping is unacceptable, then the value should be set to
`TT_Header.y_max`.

However, if a developer desires to provide appropriate default line
spacing using this field, for those applications that continue to use
this field for doing so (against OpenType recommendations), then the
value should be set appropriately. In such a case, it may result in
some glyph bitmaps being clipped.
"""

TT_OS2_win_descent = """
The descender metric for Windows. This, too, is distinct from Apple's
`TT_HoriHeader.descender` value and from the
`typo_descender`. `win_descent` is computed as the ``-y_min`` for all
characters in the Windows ANSI character set. `win_descent` is used to
compute the Windows font height and default line spacing. For platform
3 encoding 0 fonts, it is the same as ``-TTHeader.y_min``. Windows
will clip the bitmap of any portion of a glyph that appears below this
value. Some applications use this value to determine default line
spacing. This is strongly discouraged. The typographic ascender,
descender and line gap fields in conjunction with unitsPerEm should be
used for this purpose. Developers should set this field keeping the
above factors in mind.

If any clipping is unacceptable, then the value should be set to
`TT_Header.y_min`.

However, if a developer desires to provide appropriate default line
spacing using this field, for those applications that continue to use
this field for doing so (against OpenType recommendations), then the
value should be set appropriately. In such a case, it may result in
some glyph bitmaps being clipped.
"""

TT_OS2_x_height = """
This metric specifies the distance between the baseline and the
approximate height of non-ascending lowercase letters measured in font
units. This value would normally be specified by a type designer but
in situations where that is not possible, for example when a legacy
font is being converted, the value may be set equal to the top of the
unscaled and unhinted glyph bounding box of the glyph encoded at
``U+0078`` (``LATIN SMALL LETTER X``). If no glyph is encoded in this
position the field should be set to 0.

This metric, if specified, can be used in font substitution: the
`x_height` value of one font can be scaled to approximate the apparent
size of another.
"""

TT_OS2_cap_height = """
This metric specifies the distance between the baseline and the
approximate height of uppercase letters measured in font units. This
value would normally be specified by a type designer but in situations
where that is not possible, for example when a legacy font is being
converted, the value may be set equal to the top of the unscaled and
unhinted glyph bounding box of the glyph encoded at ``U+0048`` (``LATIN
CAPITAL LETTER H``). If no glyph is encoded in this position the field
should be set to 0.

This metric, if specified, can be used in systems that specify type
size by capital height measured in millimeters. It can also be used as
an alignment metric; the top of a drop capital, for instance, can be
aligned to the `cap_height` metric of the first line of text.
"""

TT_OS2_default_char = """
Whenever a request is made for a character that is not in the font,
Windows provides this default character. If the value of this field is
zero, glyph ID 0 is to be used for the default character otherwise
this is the Unicode encoding of the glyph that Windows uses as the
default character. This field cannot represent supplementary character
values (codepoints greater than ``0xFFFF``), and so applications are
strongly discouraged from using this field.
"""

TT_OS2_break_char = """
This is the Unicode encoding of the glyph that Windows uses as the
break character. The break character is used to separate words and
justify text. Most fonts specify 'space' as the break character. This
field cannot represent supplementary character values (codepoints
greater than ``0xFFFF``), and so applications are strongly discouraged
from using this field.
"""

TT_OS2_max_context = """
The maximum length of a target glyph context for any feature in this
font. For example, a font which has only a pair kerning feature should
set this field to 2. If the font also has a ligature feature in which
the glyph sequence "f f i" is substituted by the ligature "ﬃ", then
this field should be set to 3. This field could be useful to
sophisticated line-breaking engines in determining how far they should
look ahead to test whether something could change that effects the
line breaking. For chaining contextual lookups, the length of the
string (covered glyph) + (input sequence) + (lookahead sequence)
should be considered.
"""

File: src/tt_os2.c
/*
Copyright (c) 2015, Michael Droettboom
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies,
either expressed or implied, of the FreeBSD Project.
*/

#include "tt_os2.h"
#include "doc/tt_os2.h"

#include "constants.h"
#include "face.h"


#include FT_TRUETYPE_TABLES_H

#define MAKE_TT_OS2_GETTER(name, convert_func, member)  \
    MAKE_GETTER(Py_TT_OS2, name, convert_func, member)
#define TT_OS2_DEF_GETTER(name) DEF_GETTER(name, doc_TT_OS2_ ## name)


ftpy_ConstantType Py_TT_WEIGHT_CLASS_ConstantType;
ftpy_ConstantType Py_TT_WIDTH_CLASS_ConstantType;
ftpy_ConstantType Py_TT_FS_SELECTION_BitflagType;


/****************************************************************************
 Object basics
*/


typedef struct {
    ftpy_Object base;
    PyObject *owner;
    TT_OS2 *x;
} Py_TT_OS2;


static PyTypeObject Py_TT_OS2_Type;


PyObject *
Py_TT_OS2_cnew(PyObject *face)
{
    Py_Face *f = (Py_Face *)face;
    Py_TT_OS2 *self;
    TT_OS2 *orig;

    orig = (TT_OS2 *)FT_Get_Sfnt_Table(f->x, ft_sfnt_os2);
    if (orig == NULL) {
        PyErr_SetString(PyExc_AttributeError, "No os2 found");
        return NULL;
    }

    self = (Py_TT_OS2 *)(&Py_TT_OS2_Type)->tp_alloc(&Py_TT_OS2_Type, 0);
    if (self == NULL) {
        return NULL;
    }
    Py_INCREF(face);
    self->base.owner = face;
    self->x = orig;
    return (PyObject *)self;
}


static int
Py_TT_OS2_init(Py_TT_OS2 *self, PyObject *args, PyObject *kwds)
{
    PyErr_SetString(
        PyExc_RuntimeError,
        "TT_OS2 objects may not be instantiated directly. "
        "Use Face.get_os2() instead.");
    return -1;
}


/****************************************************************************
 Getters
*/

MAKE_TT_OS2_GETTER(version, PyLong_FromUnsignedLong, self->x->version)
MAKE_TT_OS2_GETTER(x_avg_char_width, PyLong_FromLong, self->x->xAvgCharWidth)


static PyObject *weight_class_get(Py_TT_OS2 *self, PyObject *closure)
{
    return Py_Constant_cnew(
        &Py_TT_WEIGHT_CLASS_ConstantType, self->x->usWeightClass);
}


static PyObject *width_class_get(Py_TT_OS2 *self, PyObject *closure)
{
    return Py_Constant_cnew(
        &Py_TT_WIDTH_CLASS_ConstantType, self->x->usWidthClass);
}


static PyObject *type_get(Py_TT_OS2 *self, PyObject *closure)
{
    return Py_Constant_cnew(
        &Py_FT_FSTYPE_BitflagType, self->x->fsType);
}


MAKE_TT_OS2_GETTER(y_subscript_x_size, PyLong_FromLong, self->x->ySubscriptXSize)
MAKE_TT_OS2_GETTER(y_subscript_y_size, PyLong_FromLong, self->x->ySubscriptYSize)
MAKE_TT_OS2_GETTER(y_subscript_x_offset, PyLong_FromLong, self->x->ySubscriptXOffset)
MAKE_TT_OS2_GETTER(y_subscript_y_offset, PyLong_FromLong, self->x->ySubscriptYOffset)
MAKE_TT_OS2_GETTER(y_superscript_x_size, PyLong_FromLong, self->x->ySuperscriptXSize)
MAKE_TT_OS2_GETTER(y_superscript_y_size, PyLong_FromLong, self->x->ySuperscriptYSize)
MAKE_TT_OS2_GETTER(y_superscript_x_offset, PyLong_FromLong, self->x->ySuperscriptXOffset)
MAKE_TT_OS2_GETTER(y_superscript_y_offset, PyLong_FromLong, self->x->ySuperscriptYOffset)
MAKE_TT_OS2_GETTER(y_strikeout_size, PyLong_FromLong, self->x->yStrikeoutSize)
MAKE_TT_OS2_GETTER(y_strikeout_position, PyLong_FromLong, self->x->yStrikeoutPosition)


/* FamilyClass and FamilySubClass seem archaic, little used and
   complicated, so they aren't wrapped. */


static PyObject *family_class_get(Py_TT_OS2 *self, PyObject *closure)
{
    return PyLong_FromUnsignedLong(
        (self->x->sFamilyClass & 0xff00) >> 8);
}


static PyObject *family_subclass_get(Py_TT_OS2 *self, PyObject *closure)
{
    return PyLong_FromUnsignedLong(
        self->x->sFamilyClass & 0xff);
}


static PyObject *panose_get(Py_TT_OS2 *self, PyObject *closure)
{
    return PyBytes_FromStringAndSize(
        (const char *)self->x->panose, 10);
}

/* unicode_range -- the specification is unfinished, so there is no
   point in getting this out. */

static PyObject *vend_id_get(Py_TT_OS2 *self, PyObject *closure)
{
    return PyUnicode_FromStringAndSize(
        (const char *)self->x->achVendID, 4);
}


static PyObject *selection_get(Py_TT_OS2 *self, PyObject *closure)
{
    return Py_Constant_cnew(
        &Py_TT_FS_SELECTION_BitflagType, self->x->fsSelection);
}


MAKE_TT_OS2_GETTER(first_char_index, PyLong_FromUnsignedLong, self->x->usFirstCharIndex);
MAKE_TT_OS2_GETTER(last_char_index, PyLong_FromUnsignedLong, self->x->usLastCharIndex);
MAKE_TT_OS2_GETTER(typo_ascender, PyLong_FromLong, self->x->sTypoAscender);
MAKE_TT_OS2_GETTER(typo_descender, PyLong_FromLong, self->x->sTypoDescender);
MAKE_TT_OS2_GETTER(typo_line_gap, PyLong_FromLong, self->x->sTypoLineGap);
MAKE_TT_OS2_GETTER(win_ascent, PyLong_FromUnsignedLong, self->x->usWinAscent);
MAKE_TT_OS2_GETTER(win_descent, PyLong_FromUnsignedLong, self->x->usWinDescent);
MAKE_TT_OS2_GETTER(x_height, PyLong_FromLong, self->x->sxHeight);
MAKE_TT_OS2_GETTER(cap_height, PyLong_FromLong, self->x->sCapHeight);
MAKE_TT_OS2_GETTER(default_char, PyLong_FromUnsignedLong, self->x->usDefaultChar);
MAKE_TT_OS2_GETTER(break_char, PyLong_FromUnsignedLong, self->x->usBreakChar);
MAKE_TT_OS2_GETTER(max_context, PyLong_FromUnsignedLong, self->x->usMaxContext);


static PyGetSetDef Py_TT_OS2_getset[] = {
    TT_OS2_DEF_GETTER(version),
    TT_OS2_DEF_GETTER(x_avg_char_width),
    TT_OS2_DEF_GETTER(weight_class),
    TT_OS2_DEF_GETTER(width_class),
    TT_OS2_DEF_GETTER(type),
    TT_OS2_DEF_GETTER(y_subscript_x_size),
    TT_OS2_DEF_GETTER(y_subscript_y_size),
    TT_OS2_DEF_GETTER(y_subscript_x_offset),
    TT_OS2_DEF_GETTER(y_subscript_y_offset),
    TT_OS2_DEF_GETTER(y_superscript_x_size),
    TT_OS2_DEF_GETTER(y_superscript_y_size),
    TT_OS2_DEF_GETTER(y_superscript_x_offset),
    TT_OS2_DEF_GETTER(y_superscript_y_offset),
    TT_OS2_DEF_GETTER(y_strikeout_size),
    TT_OS2_DEF_GETTER(y_strikeout_position),
    TT_OS2_DEF_GETTER(family_class),
    TT_OS2_DEF_GETTER(family_subclass),
    TT_OS2_DEF_GETTER(panose),
    TT_OS2_DEF_GETTER(vend_id),
    TT_OS2_DEF_GETTER(selection),
    TT_OS2_DEF_GETTER(first_char_index),
    TT_OS2_DEF_GETTER(last_char_index),
    TT_OS2_DEF_GETTER(typo_ascender),
    TT_OS2_DEF_GETTER(typo_descender),
    TT_OS2_DEF_GETTER(typo_line_gap),
    TT_OS2_DEF_GETTER(win_ascent),
    TT_OS2_DEF_GETTER(win_descent),
    TT_OS2_DEF_GETTER(x_height),
    TT_OS2_DEF_GETTER(cap_height),
    TT_OS2_DEF_GETTER(default_char),
    TT_OS2_DEF_GETTER(break_char),
    TT_OS2_DEF_GETTER(max_context),
    {NULL}
};


/****************************************************************************
 Setup
*/


static PyTypeObject Py_TT_WEIGHT_CLASS_Type;
#define TT_WEIGHT_CLASS_CONST(name) DEF_CONST(name, TT_WEIGHT_CLASS)
static constant_def TT_WEIGHT_CLASS_constants[] = {
    {"ULTRA_LIGHT", 1},
    {"EXTRA_LIGHT", 2},
    {"LIGHT", 3},
    {"SEMI_LIGHT", 4},
    {"MEDIUM", 5},
    {"NORMAL", 5},
    {"SEMI_BOLD", 6},
    {"BOLD", 7},
    {"EXTRA_BOLD", 8},
    {"ULTRA_BOLD", 9},
    {NULL}
};


static PyTypeObject Py_TT_WIDTH_CLASS_Type;
#define TT_WIDTH_CLASS_CONST(name) DEF_CONST(name, TT_WIDTH_CLASS)
static constant_def TT_WIDTH_CLASS_constants[] = {
    {"ULTRA_CONDENSED", 100},
    {"EXTRA_CONDENSED", 200},
    {"CONDENSED", 300},
    {"SEMI_CONDENSED", 400},
    {"MEDIUM", 500},
    {"NORMAL", 500},
    {"SEMI_EXPANDED", 600},
    {"EXPANDED", 700},
    {"EXTRA_EXPANDED", 800},
    {"ULTRA_EXPANDED", 900},
    {NULL}
};


static PyTypeObject Py_TT_FS_SELECTION_Type;
#define TT_FS_SELECTION_CONST(name) DEF_CONST(name, TT_FS_SELECTION)
static constant_def TT_FS_SELECTION_constants[] = {
    {"ITALIC", 0x1},
    {"UNDERSCORE", 0x2},
    {"NEGATIVE", 0x4},
    {"OUTLINED", 0x8},
    {"STRIKEOUT", 0x10},
    {"BOLD", 0x20},
    {NULL}
};


static PyTypeObject Py_TT_OS2_Type;


int setup_TT_OS2(PyObject *m)
{
    memset(&Py_TT_OS2_Type, 0, sizeof(PyTypeObject));
    Py_TT_OS2_Type = (PyTypeObject) {
        .tp_name = "freetypy.TT_OS2",
        .tp_basicsize = sizeof(Py_TT_OS2),
        .tp_doc = doc_TT_OS2__init__,
        .tp_getset = Py_TT_OS2_getset,
        .tp_init = (initproc)Py_TT_OS2_init,
    };

    ftpy_setup_type(m, &Py_TT_OS2_Type);

    if (PyType_Ready(&Py_TT_OS2_Type) < 0)
        return -1;
    Py_INCREF(&Py_TT_OS2_Type);
    PyModule_AddObject(m, "TT_OS2", (PyObject *)&Py_TT_OS2_Type);

    if (define_constant_namespace(
            m, &Py_TT_WEIGHT_CLASS_Type, &Py_TT_WEIGHT_CLASS_ConstantType,
            "freetypy.TT_WEIGHT_CLASS",
            doc_TT_WEIGHT_CLASS, TT_WEIGHT_CLASS_constants)) {
        return -1;
    }

    if (define_constant_namespace(
            m, &Py_TT_WIDTH_CLASS_Type, &Py_TT_WIDTH_CLASS_ConstantType,
            "freetypy.TT_WIDTH_CLASS",
            doc_TT_WIDTH_CLASS, TT_WIDTH_CLASS_constants)) {
        return -1;
    }

    if (define_bitflag_namespace(
            m, &Py_TT_FS_SELECTION_Type, &Py_TT_FS_SELECTION_BitflagType,
            "freetypy.TT_FS_SELECTION",
            doc_TT_FS_SELECTION, TT_FS_SELECTION_constants)) {
        return -1;
    }

    return 0;
}
